### ISSUE TITLE
Add finite assumptions to Symbol by default

### ISSUE THREAD

**Issue by oscarbenjamin** [sympy/sympy#17224]
Currently a symbol defined with no assumptions e.g. `Symbol('x')` is not assumed to be finite:
```julia
In [34]: x = Symbol('x')

In [35]: print(x.is_finite)
None
```
Pow.is_zero correctly handles this:
```julia
In [42]: m = Mul(0, x, evaluate=False)

In [43]: m
Out[43]: 0⋅x

In [44]: print(m.is_zero)
None
```
However:
```julia
In [45]: m.doit()
Out[45]: 0

In [46]: 0*x
Out[46]: 0
```
Note that if x is *explicitly* declared as infinite then we get:
```julia
In [41]: 0*Symbol('x', infinite=True)
Out[41]: nan
```
There is a contradiction here. Treating the assumptions strictly as defined `0*x` should not evaluate with default assumptions on x. That seems too painful to contemplate changing though.

It seems to me that the assumption that an arbitrary symbol is finite is baked in here and probably some other places. Rather than change that I propose to add `finite=True` by default for a Symbol. Then `0*x -> 0` can be formally correct except when someone goes out of their way to define a possibly infinite symbol.

**Comment by asmeurer** [comment]
I'd be curious if you add it as a default what breaks. I suspect this could be a very difficult thing to change (but I could be wrong).

**Comment by asmeurer** [comment]
I opened a testing PR to see what Travis tests fail https://github.com/sympy/sympy/pull/17227

**Comment by oscarbenjamin** [comment]
I tried this:
```diff
diff --git a/sympy/core/symbol.py b/sympy/core/symbol.py
index 07ff5d4..0e5da3b 100644
--- a/sympy/core/symbol.py
+++ b/sympy/core/symbol.py
@@ -231,6 +231,8 @@ def __new__(cls, name, **assumptions):
         False
 
         """
+        if not assumptions:
+            assumptions = {'finite': True}
         cls._sanitize(assumptions, cls)
         return Symbol.__xnew_cached_(cls, name, **assumptions)
```
Running the tests in core I get a few failures but they are just assumptions tests and the odd test for corner cases to do with symbols being possibly infinite e.g.:
```python
   assert Eq(1/(1/x + 1), 1).func is Eq
```
That test would be invalidated by the change (a PR could set the assumptions on x to fix it). There's a failure that I think is because this needs to be done for Dummy as well:
```python
    def test_disambiguate():
        x, y, y_1, _x, x_1, x_2 = symbols('x y y_1 _x x_1 x_2')
        t1 = Dummy('y'), _x, Dummy('x'), Dummy('x')
        t2 = Dummy('x'), Dummy('x')
        t3 = Dummy('x'), Dummy('y')
        t4 = x, Dummy('x')
        t5 = Symbol('x', integer=True), x, Symbol('x_1')
    
>       assert disambiguate(*t1) == (y, x_2, x, x_1)
E       assert (y, x_2, x, x_1) == (y, x_2, x, x_1)
E         Use -v to get the full diff
```

Running full tests now...

**Comment by oscarbenjamin** [comment]
Just saw your message. #17277 looks a better way to do it...

**Comment by oscarbenjamin** [comment]
Lots of tests using srepr fail:
```
In [2]: srepr(Symbol('x'))                                                                                                                                                                                         
Out[2]: "Symbol('x', finite=True)"
```
Of course if finite=True is the default then srepr can be adjusted for that.

Maybe needs to be fixed like for commutative:
https://github.com/sympy/sympy/blob/fa8328d1cd9e7af7c81fdef579e788038465b665/sympy/core/symbol.py#L244-L259

**Comment by hewhodares** [comment]
Moving comment over:

I think what I said was reasonable, it's not as if this topic is new ground - it /is/ unreasonable to expect me to tutor you, this is why we have lectures, I don't have to care (or pretend to, or worry about feelings) when I lecture, I don't have to worry about the rights or wrongs of each's conceptions, I just give them the correct stuff and they integrate it.

Now I am tutoring you about feelings - well outside of my comfort zone - what more of a gesture of good faith could you want?

For the extended stuff, indeed there is! Welcome to type theory, it sucks doesn't it.

I, as I am sure we all do, long for something better than paper, but here is why we don't have it, there are also variations on "extended" - for example single infinity point and stereographic projections (this comes from closure in topology)

I don't have time to go into it now, but there are strong reasons (and proofs that can be offered once we have definitions in place) to only "add" attributes to types, never say "X without Y" only "U with V") based on ending up with things that don't or can't exist.

And if you extend a types far enough you end up with something equiv to theorem provers which are also, shall we say "very hard" computationally.

**Comment by hewhodares** [comment]
I lost a line of dashes there, is there an edit button?

**Comment by hewhodares** [comment]
Forgive the triple post, want some books on this? They're often more computability-theory linked, but the problems apply. 

This is why I'm in the CAS world, I hate that I rely on paper, I hate that I know 6 or 7 CASses so well, I should warn you all now that I am just the messenger here, there is no grand fix that we just have to work together to find. 

Ultimately whatever happens will p*$$ off someone - shall we say. By dumb luck I found this and I'm urging you towards a more software-development-minded solution

**Comment by asmeurer** [comment]
Quoting @oscargus in #17227:

> I think this a valid issue and what I am primarily worried about: #17258

I responded there. I think it's a mistake to use `Eq` in the printer and expect it to always print something with `"="`, this issue notwithstanding.

> 
> Ideally, SymPy should eventually complain upon initialization. Here, the user wants to perform a summation with an integer (and therefore finite) symbol, which is the set to go from `-oo`. Sums and Integrals are typical cases where you want to put "normal" constraints on a variable, but still include `oo`. Yes, there are `extended_real` and one could define `extended_integer`, but it turns out to be quite a few `extended_*`.

Sum limits include the upper or lower bounds unless they are infinite. `Sum(f(n), (n, -oo, oo))` does not mean `-oo` and `oo` are literally included as summation terms.

> 
> I cannot really see what impact it does have in practice to make Variables finite by default. I mean, x*0 is 0 at the moment. There are the discussions about `x*(x+1) >= x**2 + x` (or if it is vice versa, I do not really understand the fundamentals).
> 
> The potential problem I see is that people do a lot of work and then let a variable be oo, instead of letting it go towards oo.

If you mean the problem of doing `expr.subs(x, oo)` instead of `limit(expr, x, oo)` this issue certainly won't solve it. You really need to perform the limit algorithm in general to get correct answers. There's no way for a "local infinity", as it were, to be aware of the global growth rate of the function it is in. 

> 
> On the other hand, I must say that setting assumptions upon initialization is a rather nice approach. Much better compared to refine expressions, provided in every call, or let variables with assumptions print differently so that they cannot be copied and pasted easily as some of the commercial alternatives.
> 
> So:
> 
> 1. How many more `extended_classes` are needed? `extended_integer` for sure. Probably `extended_imaginary` (since a `complex` may be infinite it makes sense).

I don't think we need either of these. extended integers are rare. If anything, this shows the limitation of the old assumptions, which is you can only do a union of a fixed set of assumptions. With new style assumptions you can express anything. Extended integer, if you really wanted that, would just be `Q.integer | Q.infinite`. Or if you wanted to express it as a set `Integers.union({-oo, oo})` (the new assumptions don't support sets yet but it could be added).

> 2. As variables are complex (unless otherwise stated), it is not clear how making them finite works out. I'm tempted to add `extended_complex`.

We often say variables are complex in the documentation, but it's a bit misleading. The only actual default assumption on variables is that they are commutative. `Symbol('x')` with no further assumptions could be infinite, or it could be some commutative non-number. The point of this issue is to in fact make it so that `Symbol('x')` represents a finite complex number. 

> 3. How will integrals, summations, and products with oo be handled?

I don't think it's a problem for the reason I mentioned above. Infinite limits are defined as a limit toward infinity. Infinity itself is not actually a value.

> 4. For summations, will something more happen than just replacing every is_integer with is_extended_integer?

I don't think any changes are needed. In fact, does summation even make use of is_integer now? 

> 5. Will there also be a check in subs for not doing substitutions that breaks the assumptions?

I'd say that's a separate issue. I vaguely remember it being discussed before and rejected, for some of the same reasons we allow `expr.subs(x, x + 1)`, even though x cannot "equal" x + 1 mathematically.

> 
> I think these are things to consider in the discussion.
> 
> The upside, as I see it, is that expressions can be simplified, especially relations. However, not sure if it is better to get a more complicated expression that one later on can recompute when one realizes that it is possible to set the variables to finite vs getting a simpler expression that isn't correct for infinite values...

You have to start from the least restrictive assumption from the get-go. If you make a simplification that is only correct if the value is not infinite, then by definition, you cannot then later substitute an infinite value. This is why for instance we don't assume real by default, even though many users who don't care about nonreal numbers would find it much more helpful. For instance, if you assume `x` is positive and simplify `sqrt(x**2)` to `x`, you can't later "fix" the expression to make `x` negative. You have to know which `x`s were just `x` and which were `sqrt(x**2)`. 

This is probably the strongest argument against this.

**Comment by oscarbenjamin** [comment]
Although the title is about changing the assumptions on Symbol the issue that `0*x` goes to zero when x is possibly infinite applies to other expressions as well e.g.:
```
In [12]: x = Symbol('x', real=True)                                                                                               

In [13]: 0/x                                                                                                                      
Out[13]: 0

In [14]: 0*(1/x)                                                                                                                  
Out[14]: 0
```
It seems to be baked into Mul and possibly other places that `0*expr` is zero unless expr is known to be infinite.

**Comment by asmeurer** [comment]
Indeed, 1/x would still be nonfinite if x is finite unless we also assumed nonzero. I really don't think there's an easy way around that, and making 0*x not give 0 seems like a nonstarter. I think it would really break everything.

**Comment by oscargus** [comment]
> I responded there. I think it's a mistake to use `Eq` in the printer and expect it to always print something with `"="`, this issue notwithstanding.

I agree that in this case it is a bad way of approaching printing. But it illustrates the problem to some extent.

> Sum limits include the upper or lower bounds unless they are infinite. `Sum(f(n), (n, -oo, oo))` does not mean `-oo` and `oo` are literally included as summation terms.

You are most likely mathematically right (more likely than that I am at least), but consider this example (actual SymPy test):
`Sum(KroneckerDelta(m, n), (m, -oo, oo)).doit()`
leads to
```
⎧1  for n ≥ -∞ ∧ n < ∞
⎨                     
⎩0      otherwise     
```
Now, can `n` really be `-oo`? A natural assumption, say I am a digital signal processing student, here would be to make `n` an integer, leading to that the result is 1 and (maybe) knowledge is lost. 1 is OK with me, but doesn't it somehow also lead to that `oo` effectively disappears from all discrete/integer algorithms?

(There weren't loads of these to change in #17259 but enough to notice that things changes when one, in this case, inherits assumptions based on the limits.)

> If you mean the problem of doing `expr.subs(x, oo)` instead of `limit(expr, x, oo)` this issue certainly won't solve it. You really need to perform the limit algorithm in general to get correct answers. There's no way for a "local infinity", as it were, to be aware of the global growth rate of the function it is in.

Maybe I'm more thinking about:

`Sum(KroneckerDelta(m, n), (m, -M, M)).doit()`
leading to 
```
⎧1  for M ≥ -n ∧ M ≥ n
⎨                     
⎩0      otherwise     
```
Here, one could imagine that it should be possible to subs M with oo (no assumptions made on M), leading to:
```
⎧1  for n ≤ ∞
⎨            
⎩0  otherwise
```
or simply doing a paper and pen replacement:
```
⎧1  for ∞ ≥ -n ∧ ∞ ≥ n
⎨                     
⎩0      otherwise     
```
(limit gives an error, btw).

So, depending on how we approach this we have three different answers, without a single assumption made. Yes, it is sort of a toy example and maybe it doesn't show my point, but, well, if you do not make any assumptions on stuff, then there shouldn't be any assumptions, I guess.

Oh, and with integer assumptions on `n` the result is 1.

Maybe this would have happened otherwise as well and I cannot really tell which is the correct answer, but it is a bit confusing and I'm worried that there will be more things like this.

> > 1. How many more `extended_classes` are needed? `extended_integer` for sure. Probably `extended_imaginary` (since a `complex` may be infinite it makes sense).
> 
> I don't think we need either of these. extended integers are rare. If anything, this shows the limitation of the old assumptions, which is you can only do a union of a fixed set of assumptions. With new style assumptions you can express anything. Extended integer, if you really wanted that, would just be `Q.integer | Q.infinite`. Or if you wanted to express it as a set `Integers.union({-oo, oo})` (the new assumptions don't support sets yet but it could be added).

Common enough to show up in very simple summations (at least from a programming/typing perspective). Here, ideally one would clearly like to state `n` to be `extended_integer` to not loose the additional information about possible oo values (although from some perspective one may question if these types of special cases really are relevant, I can imagine many more theoretically aimed persons thinking so).


> > 1. As variables are complex (unless otherwise stated), it is not clear how making them finite works out. I'm tempted to add `extended_complex`.
> 
> We often say variables are complex in the documentation, but it's a bit misleading. The only actual default assumption on variables is that they are commutative. `Symbol('x')` with no further assumptions could be infinite, or it could be some commutative non-number. The point of this issue is to in fact make it so that `Symbol('x')` represents a finite complex number.

At least based on the table that @oscarbenjamin suggested, `complex` in itself is not finite. The comment should be viewed from that perspective.

> > 1. How will integrals, summations, and products with oo be handled?
> 
> I don't think it's a problem for the reason I mentioned above. Infinite limits are defined as a limit toward infinity. Infinity itself is not actually a value.

No, but it is used as one and one can assign variables to it.

> 
> > 1. For summations, will something more happen than just replacing every is_integer with is_extended_integer?
> 
> I don't think any changes are needed. In fact, does summation even make use of is_integer now?

It does. Both is_integer and is_Integer. At least when it comes to determining if some simpler evaluation schemes can be used (such as direct evaluation).

> > 1. Will there also be a check in subs for not doing substitutions that breaks the assumptions?
> I'd say that's a separate issue. I vaguely remember it being discussed before and rejected, for some of the same reasons we allow `expr.subs(x, x + 1)`, even though x cannot "equal" x + 1 mathematically.

Maybe, but it will also be rather crucial when it comes to writing summations and products. I understand that from a mathematical perspective a sum from `n = -oo` doesn't mean that it takes on that "value", but from a programming perspective it does. This is a potential source of confusion, I would say.

I should also add that I enjoy the assumptions and tend to add as much assumptions on variables as I can. Still, it feels wrong to set summation variables to integer and then let them go to oo. Slightly better with the extended_real now that I have gotten used to it, but real would have also felt natural, probably more so.

**Comment by oscarbenjamin** [comment]
> `complex` in itself is not finite

AFAICT there is consensus that complex *should* be finite. There is a PR #16978 for that. There is also talk of extended complex but note that extended real and extended complex are concepts that have well-known formal definitions which is not the same as simply adding infinity to a set.

For `Sum(KroneckerDelta(m, n), (m, -oo, oo)).doit()` I think that it is meaningless to ask whether n<=oo. The `m=oo` "term" is not included in the sum in any case. Perhaps a better result would be:
```julia
In [10]: Piecewise((1, Contains(n, Integers)), (0, True))                                                                         
Out[10]: 
⎧1  for n ∈ ℤ
⎨            
⎩0  otherwise
```

I don't know whether KroneckerDelta should be restricted to integers (I have certainly never seen it used any other way outside of SymPy) but I think one reason that some calculations in SymPy end up with clauses like `x<oo` is so that substituting a non-real number raises an error. In that context it starts to make sense that defining the symbol as e.g. real leads to the condition disappearing.

**Comment by oscarbenjamin** [comment]
Evaluating a possible 0/0 can lead to incorrect results:
```julia
In [39]: n, m = symbols('n, m')                                                                                                   

In [40]: e = 1/(1+n/m)                                                                                                            

In [41]: e                                                                                                                        
Out[41]: 
  1  
─────
    n
1 + ─
    m

In [42]: e.subs(n, 0).subs(m, 0)                                                                                                  
Out[42]: 1

In [43]: e.subs(m, 0).subs(n, 0)                                                                                                  
Out[43]: nan
```
I think that the nan here is fine but the other result evaluates to 1 which is a problem.

**Comment by oscarbenjamin** [comment]
> Will there also be a check in subs for not doing substitutions that breaks the assumptions?

It would be good at least to add an optional way of checking assumptions during substitution.

**Comment by asmeurer** [comment]
> It does. Both is_integer and is_Integer. At least when it comes to determining if some simpler evaluation schemes can be used (such as direct evaluation).

Maybe we are thinking of different things. What Sum should probably do is before any calculations are done, replace the summation variables with dummies that have integer assumptions (other assumptions as well if they are warranted by the limits). Then swap them out at the end of the calculation. If that happened you wouldn't get the `n < oo` expression you mentioned, because it would go away from the assumptions on the dummy. See also https://github.com/sympy/sympy/pull/17093.

> Evaluating a possible 0/0 can lead to incorrect results:

It would be nice if `limit(e, (n, m), (0, 0))` were something that worked.

**Comment by oscarbenjamin** [comment]
> replace the summation variables with dummies that have integer assumptions

Maybe one day we can use refine for this:
```python
summand = refine(summand, Q.integer(n) & Q(n >= a) & Q(n <= b))
```


> It would be nice if `limit(e, (n, m), (0, 0))` were something that worked.

It's not well defined without knowing any relationship between n and m:
```julia
In [98]: (1/(1+n/m)).subs(n, x).subs(m, x**2).limit(x, 0)                                                                         
Out[98]: 0

In [99]: (1/(1+n/m)).subs(n, x**2).subs(m, x).limit(x, 0)                                                                         
Out[99]: 1
```

**Comment by asmeurer** [comment]
I think there were some examples in integral issue showing that you really need to be able to make the assumption going into the algorithm. It may not be enough to just let it handle the general case and refine later, because the algorithm might not actually have an answer for the general case. There can be simplifications that are only true for integers, for example, `(-1)**n` is a much simpler expression when `n` is an integer than when it is a general real or complex number. Most summation algorithms already implicitly assume this which is why it hasn't really been an issue, but I imagine there could be cases where they call out to other parts of SymPy that don't implicitly make this assumption.

So you need to be able to either 

```py
expr = expr.subs(symbol, dummy_with_assumptions)
result = do summation/integration/etc. algorithm on expr
result = result.subs(dummy_with_assumptions, symbol)
```

or

```py
with assuming(symbol, assumptions):
    return summation_algorithm(expr)
```

**Comment by smichr** [comment]
1. I gave an attempt at this with my `xfin` branch. My initial feeling is that this is possible and should be done. I suspect just making the default symbol be complex (-> finite) and commutative is a good way to go. 
2. printing can be fixed to only print `{'commutative': True, 'complex': True}` and to exclude the redundant `'finite': True, 'infinite': False`
3. And I think we should make `0*x` only be 0 if x is finite. And since `1/x` is not necessarily finite if `x` is finite,  `0/x` would remain unevaluated (#18366)
